cpp/tests/io/json_whitespace_normalization_test.cu

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <io/fst/lookup_tables.cuh>
+#include <io/utilities/hostdevice_vector.hpp>
+
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/testing_main.hpp>
+
+#include <rmm/cuda_stream.hpp>
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/iterator/discard_iterator.h>
+
+#include <cstdlib>
+#include <string>
+
+namespace {
+// Type used to represent the atomic symbol type used within the finite-state machine
+using SymbolT = char;
+using StateT  = char;
+
+// Type sufficiently large to index symbols within the input and output (may be unsigned)
+using SymbolOffsetT = uint32_t;
+
+enum class dfa_states : StateT { TT_OOS = 0U, TT_DQS, TT_DEC, TT_WS, TT_NUM_STATES };
+enum class dfa_symbol_group_id : uint32_t {
+  DOUBLE_QUOTE_CHAR,   ///< Quote character SG: "
+  ESCAPE_CHAR,         ///< Escape character SG: '\\'
+  NEWLINE_CHAR,        ///< Newline character SG: '\n'
+  WHITESPACE_SYMBOLS,  ///< Whitespace characters SG: '\t' or ' '
+  OTHER_SYMBOLS,       ///< SG implicitly matching all other characters
+  NUM_SYMBOL_GROUPS    ///< Total number of symbol groups
+};
+
+// Aliases for readability of the transition table
+constexpr auto TT_OOS            = dfa_states::TT_OOS;
+constexpr auto TT_DQS            = dfa_states::TT_DQS;
+constexpr auto TT_DEC            = dfa_states::TT_DEC;
+constexpr auto TT_WS             = dfa_states::TT_WS;
+constexpr auto TT_NUM_STATES     = static_cast<StateT>(dfa_states::TT_NUM_STATES);
+constexpr auto NUM_SYMBOL_GROUPS = static_cast<uint32_t>(dfa_symbol_group_id::NUM_SYMBOL_GROUPS);
+
+// The i-th string representing all the characters of a symbol group
+std::array<std::vector<SymbolT>, NUM_SYMBOL_GROUPS - 1> const wna_sgs{
```

Comment:
My assumption is that the state resets on \n, so invalid rows (i.e. mismatched quote chars) don't impact the rest of the file. Is this correct?

Reply:
Yes, the state transitions on `\n`, irrespective of whether the character is within a string or not (modified transition table in commit [595a71](https://github.com/rapidsai/cudf/pull/14931/commits/5957a71452dc5815f6b3300788827952b18f58ad) to handle the recovery of JSON lines with invalid lines).  
In the case of invalid input such as `{"a": "b }`, I'm unsure about how Spark parses this - if the whitespace before the closing brace would be retained or not. The current FST will consider it to be a part of the string and retain it, but if the alternative behavior is expected, that would change the FST.

Reply:
Could we add a test for this? For example, input `{"a": "b }\n{ "c" : "d" }` and verify that the output contains {"c":"d"}.

Reply:
I think the test `GroundTruth_MultilineJSONWithSpacesAndTabs` covers this scenario.

Reply:
Is it this line? `{ \"foo rapids\": { \"a\": 1 },` I'd like to see a test that shows that "in quotes" state does not propagate past the new line, but it's not really a blocking requirement.

Reply:
Ah okay, I understand what you're saying. I'll add the test.

Reply:
The test `GroundTruth_InvalidInput` test considers the above example.